### PR TITLE
commit-3 (updated) of multi-threaded decoding support.

### PR DESCRIPTION
**Pull request by xiaotianshi2**
Mainly update DecodeCurrentAccessUnit and associated functions to do threaded parallel frame decoding.

Add additional function PrefetchLastPicForThread and variables for better control of threaded decoding in the case of multi-slice with one slice group contains one slice.

### CODE REVIEW

File: codec/decoder/core/src/decode_slice.cpp [hunk 1]
```diff
@@ -1736,9 +1736,8 @@ int32_t WelsDecodeAndConstructSlice (PWelsDecoderContext pCtx) {
     if (!pCurDqLayer->pMbCorrectlyDecodedFlag[iNextMbXyIndex]) { //already con-ed, overwrite
       pCurDqLayer->pMbCorrectlyDecodedFlag[iNextMbXyIndex] = true;
       pCtx->pDec->iMbEcedPropNum += (pCurDqLayer->pMbRefConcealedFlag[iNextMbXyIndex] ? 1 : 0);
-      ++pCtx->iTotalNumMbRec;
     }
-
+    ++pCtx->iTotalNumMbRec;
```
[the remark this answers is not in the document]
Comment (xiaotianshi2) [reply codec/decoder/core/src/decode_slice.cpp]: Could you describe In what kind of cases in which the same MB will be decoded more than once? I'll see if I can find a better way.
In threaded mode with multi-slice, It is hard to get pMbCorrectlyDecodedFlag update correctly among threaded CTX unless I move array pMbCorrectlyDecodedFlag to where all threads can share. That's why I did this.

File: codec/decoder/core/src/decoder_core.cpp [hunk 2]
```diff
@@ -2721,34 +2826,49 @@ int32_t DecodeCurrentAccessUnit (PWelsDecoderContext pCtx, uint8_t** ppDst, SBuf
         }
       }
 
+      if (pThreadCtx != NULL && pCtx->uiDecodingTimeStamp > 1 && pCtx->pLastDecPicInfo->uiDecodingTimeStamp > 0) {
```
[the remark this answers is not in the document]
Comment (xiaotianshi2) [reply codec/decoder/core/src/decoder_core.cpp]: uiDecodingTimeStamp is an internal variable which represents the current access unit (a slice or frame) number with first access unit = 1. 0 is its initialization value.

File: codec/decoder/core/src/decode_slice.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (huili2) [reply codec/decoder/core/src/decode_slice.cpp]: an example is to send a duplicated NAL into decoder. The current decoder should support this. In this PR, I don't think it is the case.

[the remark this answers is not in the document]
Comment (xiaotianshi2) [reply codec/decoder/core/src/decode_slice.cpp]: I have updated the code to support duplicated NAL in threaded mode.

**Comment by xiaotianshi2** [comment]
@huili2 @sijchen  Can you check why it still shows that "1 pending check" even build is successful?

**Comment by huili2** [comment]
Not sure what happened. Seems all tasks are OK.
Let me restart the travis test.

File: codec/decoder/core/src/wels_decoder_thread.cpp [hunk 3]
```diff
@@ -57,6 +57,12 @@
 #define HW_NCPU_NAME "hw.ncpu"
 #endif
 #endif
+#ifdef ANDROID_NDK
+#include <cpu-features.h>
+#endif
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
```
Comment (mstorsjo) [inline codec/decoder/core/src/wels_decoder_thread.cpp]: Why do you readd these includes here? This looks like an accident.

Reviewers, please try to be careful and look for these kind of accidental changes.

This was intentionally removed in #3183, as the include path for cpu-features.h isn't available when building codec/decoder/*, only when building codec/common/*, and there's no need to actually use these functions here. Please see #3192 where I fix building for android again.

**Pull request merged**